### DISCUSSION TITLE
How to mock container resolved class with parameters?

### DISCUSSION THREAD

**Discussion by catabozan** [laravel/framework#41408] [Q&A]
I want to mock a class that is created using a static method 
```php
// this is how I instanciate the object
MyClass::from(AnotherClass $data);

// this is how the class looks like
class MyClass 
{
	 protected function __construct(AnotherClass $data)
	{
		// do stuff
	}

	 public static function from(AnotherClass $data): self
    {
        return new self($data);
    }
}
```
This is how I bind it to the container:
```php
$this->app->bind(MyClass::class, function ($app, array $parameters) {
	return MyClass::from($parameters[0]);
});
```
This is how I use the class
```php
$data = new AnotherClass()
app(MyClass::class, [$data])
```
I cannot store the `$data` globally to get rid of the parameter in my binding.

The problem is that when I try to mock or spy on `MyClass` it does not work.
```php
$spy = spy(MyClass::class);
// ... do some stuff
$spy->shouldHaveReceived('reactPHPIdle');
```

If I dump `app(MyClass::class)` the mocked class is returned. But when I dump `app(MyClass::class, [$data])` the original class is returned, not the mocked one.

How can I successfully mock my class? Is there a better way to do this (ex. refactor the class)?

**Comment by gguney** [comment]
Did you try documentation way ?

```
use App\Service;
use Mockery\MockInterface;
 
$mock = $this->mock(MyClass::class, function (MockInterface $mock) {
    $mock->shouldReceive('reactPHPIdle')->once();
});
```

**Reply by catabozan** [reply]
Yes, It didn't work. As I understand, you cannot mock a container bound class that receives parameters.
